src/internal/components/InternalThemeEditor.tsx

```diff
@@ -0,0 +1,102 @@
+import { Puck, Config, InitialHistory } from "@measured/puck";
+import React from "react";
+import { useState } from "react";
+import { TemplateMetadata } from "../types/templateMetadata.ts";
+import { EntityFieldProvider } from "../../components/EntityField.tsx";
+import { DevLogger } from "../../utils/devLogger.ts";
+import ThemeSidebar from "../puck/components/ThemeSidebar.tsx";
+import { ThemeConfig } from "../../utils/themeResolver.ts";
+import { ThemeSaveState } from "../types/themeSaveState.ts";
+import { ThemeHeader } from "../puck/components/ThemeHeader.tsx";
+
+interface InternalThemeEditorProps {
+  puckConfig: Config;
+  puckInitialHistory: InitialHistory | undefined;
+  isLoading: boolean;
+  templateMetadata: TemplateMetadata;
+  publishThemeConfiguration: (data: any) => void;
+  devLogger: DevLogger;
+  themeConfig?: ThemeConfig;
+  saveThemeSaveState: (data: any) => void;
+  themeHistory?: ThemeSaveState;
+  setThemeHistory: (themeHistory: ThemeSaveState) => void;
+  clearThemeHistory: () => void;
+}
+
+// Render Puck editor
+export const InternalThemeEditor = ({
+  puckConfig,
+  puckInitialHistory,
+  isLoading,
+  templateMetadata,
+  publishThemeConfiguration,
+  devLogger,
+  themeConfig,
+  saveThemeSaveState,
+  themeHistory,
+  setThemeHistory,
+  clearThemeHistory,
+}: InternalThemeEditorProps) => {
+  const [canEdit, setCanEdit] = useState<boolean>(false); // helps sync puck preview and save state
```

Comment:
I wonder if your iframe observer would allows us to not use this now

Reply:
When I looked at removing this a few weeks ago, it was a problem anytime you moved a component around. Probably worth looking at again while we're doing this refactoring work